android/src/main/java/com/ahm/capacitor/camera/preview/Preview.java

```diff
@@ -314,20 +319,85 @@ private Camera.Size getOptimalPreviewSize(List<Camera.Size> sizes, int w, int h)
         return optimalSize;
     }
 
-    public void surfaceChanged(SurfaceHolder holder, int format, int w, int h) {
-        if (mCamera != null) {
-            try {
-                // Now that the size is known, set up the camera parameters and begin
-                // the preview.
-                mSupportedPreviewSizes = mCamera.getParameters().getSupportedPreviewSizes();
-                if (mSupportedPreviewSizes != null) {
-                    mPreviewSize = getOptimalPreviewSize(mSupportedPreviewSizes, w, h);
+    private boolean tryFindAlternativePreviewSizes(int w, int h, Camera.Size previousSize)
+    {
+        final double ASPECT_TOLERANCE = 0.1;
+        double targetRatio = computeTargetRatio(w, h);
+
+        if (mSupportedPreviewSizes == null) {
+            return false;
+        }
+
+        Camera.Size optimalSize = null;
+        double minDiff = Double.MAX_VALUE;
+
+        int targetHeight = h;
+
+        boolean valueAlreadyConsidered = true;
+        // Try to find an size match aspect ratio and size
+        for (Camera.Size size : mSupportedPreviewSizes) {
+            if(valueAlreadyConsidered) {
+                if(size.width == previousSize.width && size.height == previousSize.height) {
+                    valueAlreadyConsidered = false;
                 }
-                startCamera();
-            } catch (Exception exception) {
-                Log.e(TAG, "Exception caused by surfaceChanged()", exception);
+                continue;
+            }
+
+            double ratio = (double) size.width / size.height;
+            if (Math.abs(ratio - targetRatio) > ASPECT_TOLERANCE) continue;
+            if (Math.abs(size.height - targetHeight) < minDiff) {
+                optimalSize = size;
+                minDiff = Math.abs(size.height - targetHeight);
+            }
+        }
+
+        if(optimalSize == null) {
+            return false;
+        }
+
+        Log.d(TAG, "optimal preview size: w: " + optimalSize.width + " h: " + optimalSize.height);
+        mPreviewSize = optimalSize;
+        return true;
+    }
+
+    public void surfaceChanged(SurfaceHolder holder, int format, int w, int h) {
+        Camera.Parameters parameters = mCamera.getParameters();
+        mSupportedPreviewSizes = mCamera.getParameters().getSupportedPreviewSizes();
+        if (mSupportedPreviewSizes == null) {
+            restartCamera();
+            return;
+        }
+
+        mPreviewSize = getOptimalPreviewSize(mSupportedPreviewSizes, w, h);
+        boolean success = trySetPreviewSize(parameters, mPreviewSize);
+
+        while(!success) {
+            if(tryFindAlternativePreviewSizes(w, h, mPreviewSize)) {
```

Comment:
Terminiert das hier denn sicher, wenn du immer nur den letzten Candidate übergibst und ausschließt? 🤔 

Bsp.:
In Run 1 kriegst du 200x200 zurück, beim Setzen gibt's aber eine Exception.
In Run 2 schließt du 200x200 aus, und kriegst 100x100 zurück. Es gibt wieder eine Exception.
In Run 3 schließst du 100x100 aus, und kriegst wieder 200x200 => Exception. Repeat.


edit: Ah, jetzt versteh ich. In der `tryFindAlternativePreviewSizes` setzt du dein ignore gleich am Anfang auf true und skippst so lange, bis du einen neuen Wert hast. Das setzt aber im ersten Schritt voraus, dass `mPreviewSize` == `mSupportedPreviewSizes[0]`, ist das denn garantiert so?

Ich fänds leichter verständlich wenn du hier lokal ein Array hättest, wo du jeden versuchten Candidate reingibst und dass dann zum Ausschließen verwendest

Reply:
`mPreviewSize` ist im ersten Schritt der Wert der bei `getOptimalPreviewSize` rauskommt. Also der erste in der Liste der der gesuchten aspect ratio am besten entspricht. (Wobei mir gerade aufgefallen ist, dass das für unsere Anforderungen eventuell auch nicht optimal gelöst ist weil die Funktion auch einen Wert zurückgeben kann der davon abweicht - siehe zweite for-loop. Dann hätte man wieder eine Verzerrung).

Array ist deshalb schwierig weil eben nicht jeder Wert in Frage kommt und bei jedem Schritt welche ausgelassen werden können. Aber ja, besonders gut lesbar ist das ganze nicht, mal schauen was sich da machen lässt.

Reply:
Das geht potentiell alles so viel einfacher.. der code könnte echt ein bisschen refactoring vertragen :face_with_spiral_eyes: 

    private boolean isSizeMatchingAspectRatio(int w, int h, Camera.Size size)
    {
        final double ASPECT_TOLERANCE = 0.1;
        double targetRatio = computeTargetRatio(w, h);
        double ratio = (double) size.width / size.height;

        return (Math.abs(ratio - targetRatio) < ASPECT_TOLERANCE);
    }

    public void surfaceChanged(SurfaceHolder holder, int format, int w, int h) {
        Camera.Parameters parameters = mCamera.getParameters();
        mSupportedPreviewSizes = mCamera.getParameters().getSupportedPreviewSizes();
        if (mSupportedPreviewSizes == null) {
            restartCamera();
            return;
        }

        for(Camera.Size size : mSupportedPreviewSizes) {
            if(isSizeMatchingAspectRatio(w, h, size)) {
                mPreviewSize = size;
                if (trySetPreviewSize(parameters, mPreviewSize)) {
                    mCamera.startPreview();
                    return;
                }
            }
        }

        restartCamera();
    }
    
Muss das dann am Montag aber noch mit den jeweiligen devices testen..